server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -266,19 +264,6 @@ protected void indexScriptValues(SearchLookup searchLookup, LeafReaderContext re
         throw new UnsupportedOperationException("FieldMapper " + name() + " does not support [script]");
     }
 
-    protected final void createFieldNamesField(ParseContext context) {
-        assert fieldType().hasDocValues() == false : "_field_names should only be used when doc_values are turned off";
```

Comment:
is there a way to keep this assertion? I vaguely remember adding it while cleaning up some inconsistencies around using field_names where it would not make sense

Reply:
I've added an assertion in FieldNamesFieldMapper.postParse()